src/cli-utils.js

```diff
@@ -0,0 +1,113 @@
+/*
+ * @flow
+ */
+
+import type {BuildSpec, Config} from './types';
+import chalk from 'chalk';
+import * as fs from './lib/fs';
+import * as path from './lib/path';
+
+export type FormatBuildSpecTreeCtx = {
+  level: number,
+  isSeen: boolean,
+  isLast: boolean,
+};
+
+export type FormatBuildSpecQueueItem = {
+  spec: BuildSpec,
+  ctx: FormatBuildSpecTreeCtx,
+  path: string,
+};
+
+export function getBuildSpecPackages(
+  config: Config<*>,
+  root: BuildSpec,
+  includeDeps?: boolean = true,
+  includeSeen?: boolean = true,
+): FormatBuildSpecQueueItem[] {
+  const specsQueue = [];
+  const seen = new Set();
+
+  const queue = [
+    {
+      spec: root,
+      ctx: {
+        level: 0,
+        isSeen: false,
+        isLast: true,
+      },
+      path: '#',
+    },
+  ];
+
+  while (queue.length > 0) {
+    const cur = queue.shift();
+    const {spec, ctx, path} = cur;
+
+    const isSeen = seen.has(spec.id);
+    if (isSeen && !includeSeen) {
+      continue;
+    }
+    const {level} = ctx;
+
+    specsQueue.push({spec, ctx: {...ctx, isSeen}, path});
+    seen.add(spec.id);
+
+    if (includeDeps || level < 1) {
+      const deps = Array.from(spec.dependencies.values());
+      let idx = 0;
+      for (const dep of deps) {
+        const isLast = dep === deps[deps.length - 1];
+        queue.push({
+          spec: dep,
+          ctx: {level: level + 1, isSeen: false, isLast},
+          path: `${path}${idx++}${dep.name}#`,
```

Comment:
@andreypopp Please review this chunk. I find it clunky, but I didn't want to rebuild the tree again. 😩

Reply:
I think having intermediate tree structure for representing tree visually is ok. But let's keep it as-is now. We can refactor it later if needed.